cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java

```diff
@@ -192,48 +192,53 @@ private RouteDestination getV3RoutingService(String [] uriParts, AllowedMethod r
   }
 
   /**
-   * Determines if actual matches expected.
+   * Determines if the beginning of an array of strings matches an expected sequence of strings.
    *
-   * - actual may be longer than expected, but we'll return true as long as expected was found
-   * - null in expected means "accept any string"
+   * <ul><li>
+   *   the actual sequence may be longer than the expected one as long as its beginning matches;
+   * </li><li>
+   *   a null in the expected sequence means "accept any string" in that position.
+   * </li></ul>
    *
-   * @param actual actual string array to check
-   * @param expected expected string array format
-   * @return true if actual matches expected
+   * @param actual the actual string array to check; must not contain nulls.
+   * @param expected the expected string array to match; may contain nulls as wildcards.
+   * @return true if the start of {@code actual} matches {@code expected}
    */
-  private boolean matches(String[] actual, String... expected) {
+  @VisibleForTesting
+  static boolean beginsWith(String[] actual, String... expected) {
     if (actual.length < expected.length) {
       return false;
     }
 
     for (int i = 0; i < expected.length; i++) {
-      if (expected[i] == null) {
-        continue;
-      }
-      if (!expected[i].equals(actual[i])) {
+      if (expected[i] != null && !expected[i].equals(actual[i])) {
         return false;
       }
     }
     return true;
   }
 
   /**
-   * Determines if the actual ends with expected. If expected contains '?' at the end the the last element of actual
-   * is ignored as ? is considered to match one element regardless of what it is
+   * Determines if the end of an array of strings matches an expected sequence of strings.
+   *
+   * <ul><li>
+   *   the actual sequence may be longer than the expected one as long as its end matches;
+   * </li><li>
+   *   a null in the expected sequence means "accept any string" in that position.
+   * </li></ul>
+   *
+   * @param actual the actual string array to check; must not contain nulls.
+   * @param expected the expected string array to match; may contain nulls as wildcards.
+   * @return true if the end of {@code actual} matches {@code expected}
    */
   @VisibleForTesting
-  boolean endsWith(String[] actual, String... expectedEnd) {
-    if (expectedEnd.length > actual.length) {
+  static boolean endsWith(String[] actual, String... expected) {
+    if (actual.length < expected.length) {
       return false;
     }
-    int offset = 1;
-    if (expectedEnd[expectedEnd.length - offset].equals("?")) {
-      offset++;
-    }
-    while (expectedEnd.length - offset >= 0) {
-      if (expectedEnd[expectedEnd.length - offset].equalsIgnoreCase(actual[actual.length - offset])) {
-        offset++;
-      } else {
+    int offset = actual.length - expected.length;
+    for (int i = 0; i < expected.length; i++) {
```

Comment:
The beginsWith and endsWith seems very similar how about extracting the logic something like this maybe:

```
  static boolean beginsWith(String[] actual, String... expected) {
    return match(actual, expected, 0);
  }

  static boolean endsWith(String[] actual, String... expected) {
    return match(actual, expected, actual.length - expected.length);
  }

  private static boolean match(String[] actual, String[] expected, int offset) {
    if (actual.length < expected.length) {
      return false;
    }
    for (int i = 0; i < expected.length; i++) {
      if (expected[i] != null && !expected[i].equals(actual[offset + i])) {
        return false;
      }
    }
    return true;
  }
```

Reply:
even better :) I changed it slightly to make it a little clearer

Reply:
💯